### ISSUE TITLE
Create a Homebrew package for OSX

### ISSUE THREAD

**Issue by matthieudelaro** [matthieudelaro/nut#14]
So that OSX users can install Nut with `brew install nut`

**Comment by RnbWd** [comment]
I've dabbled with porting similar projects to brew and npm, but only if there's versioned released. I know it's possible to use the HEAD branch, but I don't know how to do that. Home-brew likes everything versioned and has guidelines etc., npm doesn't have any rules like that. All I need a tag or release - to grab the `tar.gz`. I just discovered this project 10 minutes I think it's really cool, still figuring out how it works

**Comment by RnbWd** [comment]
i just found your tar.gz and it's easy to link actually - i think somebody has the 'nut' namespace, unless it's this library, I've never used docker in a brew but it should work (I'm assuming docker mac beta is already running on the system)

**Comment by RnbWd** [comment]
I don't fully understand brew - there's lots of env variables and processes happening in a different space, and the paths might need to be defined explicitly for better compatibility with docker beta's volumes. I tried scripting it / using the brew syntax, and the scripts work locally but not in the context of brew

**Comment by matthieudelaro** [comment]
Thanks for working on this issue! I don't know about how brew works internally, but if you have any question about Nut to make it work, I'll do my best:
- Yes, you are right: assume docker for mac is already here. Also, we don't have to bother about `nvidia-docker-plugin`, since Nut can use it only on Linux (to mount graphic cards in the container). 
- the brew formula called `nut` has nothing to do with this project :( So I think we should create `docker-nut` formula.

**Comment by RnbWd** [comment]
Unless we can generate some sort of binary I have no idea how to get this brew - I don't know if anyone's created a brew that runs docker during the install process? Someone who understands how brew works and native volumes in docker could figure it out. But I could probably do it with node. 'npm install -g nut' - no esoteric APIs or anything I could probably just run the commands and it'd be symlinked into the path automatically just because it's installed with npm -g

**Comment by matthieudelaro** [comment]
Travis might be the solution: looks like we can upload files to github from the build process thanks to [GitHub Releases Uploading](https://docs.travis-ci.com/user/deployment/releases)

I've set up a [basic travis](https://travis-ci.org/matthieudelaro/nut/branches) configuration. Hopefully adding a few lines to `.travis.yml` will generate binaries automatically for us.

**Comment by matthieudelaro** [comment]
Short term solution: I created the branch [manualbuild](https://github.com/matthieudelaro/nut/tree/manualbuild), with the binaries in the release folder: https://github.com/matthieudelaro/nut/tree/manualbuild/release

**Comment by RnbWd** [comment]
i got it to download but it's saying i don't have permission to run - I haven't test the binary alone yet

**Comment by RnbWd** [comment]
yeah it works identical either way - brew is downloading the the correct binary into the path, but I can't get the file you hosted to work on my machine.

**Comment by RnbWd** [comment]
thanks for working on this - I really want to get it working with a browser gui like jess/chromium or jess/firefox - reading over the docs it looks like docker is exposing more files in a custom path /Mac - at least during the beta tests. I don't even know how to debug the docker - gui links (sorry this post is off topic)

**Comment by matthieudelaro** [comment]
When you compile Nut by yourself, does it run properly? What version of Docker do you use? Is it Docker for Mac? Toolbox?

**Comment by RnbWd** [comment]
nut works fine when i compile it locally - docker for mac - the beta - but the binary file from github had permission issues.. maybe i just need to `chmod +x` it?

**Comment by RnbWd** [comment]
yep... answered my own question lol it works now

**Comment by RnbWd** [comment]
with brew though - it also doesn't have correct permission, and I'm not sure if that can be bypassed without prompting the user to input their password during the install.. because docker for mac (and nut) need root level access. When compiling normally it doesn't matter because dockers already has those permissions, so people would have to trust that the binary actually is what it says it is (of course I know that's true).. brews documentation is terrible - but I'm sure theres a way to do it.

**Comment by RnbWd** [comment]
Also I could be wrong about that - I'm just looking at various brews in the repo and trying my best to link a binary, I've attempted for another project as well, and even if we get it working, there's two choices: 

1) submit pull request and try to stick by the 'rules' of home-brew
2) Just have users tap a repo on github and manage however you'd like

I have strong feeling that once docker native is mature, there's going to be a tap for docker-specific brews

**Comment by matthieudelaro** [comment]
Ok. I wrote a wiki explaining how Nut works: https://github.com/matthieudelaro/nut/wiki/Nut:-Under-The-Hood
Tell me if it doesn't answer to your questions. 

There are brew packages for docker:
```
$ brew search docker

boot2docker                  docker-cloud                 docker-gen                   docker-machine-driver-xhyve  docker-swarm               
docker                       docker-compose               docker-machine               docker-machine-parallels
```

So maybe those are good examples. Otherwise npm is a good start. Thanks again for your work on this.

**Comment by RnbWd** [comment]
Thanks I'll read over the internals of nut! Docker-Beta is actually an application, not a binary, so it would be downloaded with `brew cask`. The app symlinks those docker binaries into the usr/local/bin and does some other magic when you install it that requires admin privileges (osxfs and hypervisor). 

Linking a prebuilt binary is the simplest way to get it working, and it's what I'm most familiar with. When I use docker beta with home-brew - I don't know what's going on with the permissions and filesystem when both are interacting with one another? (home-brew uses different folders during install and then symlinks everything into `/usr/local/bin`, and then docker's osxfs is still being modified and sometimes files get lost in translation).

I'm not saying it's impossible, I just don't know what i'm doing. And the binary could work.. I need to read up on the documentation a little more.

**Comment by RnbWd** [comment]
home-brew is has some terrible documentation... this is why 'sudo is bad':

> tl;dr Sudo is dangerous, and you installed TextMate.app without sudo anyway.
> 
> Homebrew is designed to work without using sudo. You can decide to use it but we strongly recommend not to do so. If you have used sudo and run into a bug then it is likely to be the cause. Please don’t file a bug report unless you can reproduce it after reinstalling Homebrew from scratch without using sudo.
> 
> You should only ever sudo a tool you trust. Of course, you can trust Homebrew ;) But do you trust the multi-megabyte Makefile that Homebrew runs? Developers often understand C++ far better than they understand make syntax. It’s too high a risk to sudo such stuff. It could break your base system, or alter it subtly.
> 
> And indeed, we’ve seen some build scripts try to modify /usr even when the prefix was specified as something else entirely.
> 
> Did you chown root /Applications/TextMate.app? Probably not. So is it that important to chown root wget?
> 
> If you need to run Homebrew in a multi-user environment, consider creating a separate user account especially for use of Homebrew.
> 

I assume sudo is necessary for nut to bind with `unix:///var/run/docker.sock`?? I can easily run a script in bash without sudo privileges to create nut on a machine with docker already installed - that's how I expect brew to work, so I'm brainstorming both fronts, or maybe I'lll just use node.

**Comment by RnbWd** [comment]
I wrote a little npm script here: https://github.com/RnbWd/nut-bin (sorry all the namespaces are taken).

but it links to the binary you created a few days ago, 

`npm install -g nut-bin` 

that should link the binary into the path as **nut** (not nut-bin even though that's name on npm)

**Comment by matthieudelaro** [comment]
Nice :)

**Comment by RnbWd** [comment]
Just added more documentation, a test, and the linux build (so npm is synced with github releases) Windows is possible but I don't have a way to test it and i thought they ended with .exe? 

If producing binaries isn't an issue that repo will always link to the latest build by default -But if we're linking binaries it'd be easier if they had the same name, maybe different builds separated by folders? Mostly just to separate windows from linux / darwin but also it expects the same filename for the systems. 

I could also experiment with running the docker build in node using an exec process - i know that's possible as well, but it would take a long time to finish downloading relative to a prebuilt binary.

**Comment by matthieudelaro** [comment]
`.exe` : I think Go adds the .exe automatically if you don't specify the name of the binary when you build (which I do)

I updated the version of the build, and created a subfolder per OS in [release](https://github.com/matthieudelaro/nut/tree/manualbuild/release), with the executable with the same file name in each folder. (Linux binary and OSX binary are different). I left the previous binaries as well, so that your script works until you make use of the new files.

**Comment by RnbWd** [comment]
Thanks, I'll update the library right now! That simplifies everything.

Here's an example from 'bin-wrapper' library I'm using:

````
var base = 'https://github.com/imagemin/gifsicle-bin/raw/master/vendor';
var bin = new BinWrapper()
    .src(base + '/osx/gifsicle', 'darwin')
    .src(base + '/linux/x64/gifsicle', 'linux', 'x64')
    .src(base + '/win/x64/gifsicle.exe', 'win32', 'x64')
    .dest(path.join('vendor'))
    .use(process.platform === 'win32' ? 'gifsicle.exe' : 'gifsicle')
    .version('>=1.71');
````

 if everything is named `nut`, the  `.use()` method is  simplified without the `? statement`, so hopefully it works on all platforms when I change that line.

I'm also assume these binaries are built for x64? Just wanted to make sure I'm linking the correct architectures, but hopefully the test would fail if it wasn't the case. It might be a good idea to set up travis in the npm repo as well.

**Comment by RnbWd** [comment]
Seems to be working.  Right now  [nut-bin](https://github.com/RnbWd/nut-bin) is synced with nut's versioning scheme on npm/github and tested before being published, so it should be fairly simple to add a hook that automates the entire process. I haven't attempted homebrew again, but I'm willing to give it a second go.

**Comment by matthieudelaro** [comment]
Well done!

So far the release is compiled for 64bits only. x85, arm, ... could be added: compiling for other architectures is as easy as changing GOARCH argument before compiling (as show in the [nut.yml](https://github.com/matthieudelaro/nut/blob/master/nut.yml#L21)) (I'm amazed by the simplicity of cross-compilation of Go)

I'll remove the builds at the root of release folder when you confirm that you don't need them at all anymore.

**Comment by RnbWd** [comment]
okay I kinda messed up with linux (misspelled nut) but I put it in repo's documentation to use `npm install -g nut@1.2.0-latest` - right now i'm linking directly to the binaries but I haven't attempted using a self-install process yet (slightly different setup)

**Comment by RnbWd** [comment]
actually didn't have to change anything except for the version number, do you want me to add you onto the repo? Just in case you want to update and I'm not around, although it might just work without any changes if the url to the binary files remains constant. i use this module to publish https://github.com/sindresorhus/np

**Comment by matthieudelaro** [comment]
Ok. And yes please, that would be nice.

**Comment by RnbWd** [comment]
It's not throwing errors on windows, let me know if it works now

**Comment by matthieudelaro** [comment]
`npm install -g nut-bin` => test fails, but `nut` works anyway :)

![test_failed](https://cloud.githubusercontent.com/assets/2452725/15766312/ff25a28c-2977-11e6-8df9-f7bc7c17135a.png)

![image](https://cloud.githubusercontent.com/assets/2452725/15766319/120e1c12-2978-11e6-8d1e-fa0586c7f651.png)

**Comment by RnbWd** [comment]
I think it's because windows requires an ext in the binary - so `nut` works fine in linux / darwin - but I had to rename the file `nut.exe` - which happens after the test.. but if you could rename the file in github to `nut.exe` then I think it'll fix the tests and everything. Windows :rage1:

**Comment by RnbWd** [comment]
nice i see it's already updated

**Comment by RnbWd** [comment]
seems to work now - I forgot to push the updates last night onto github, but it doesn't matter because `nut.exe` fixed all the issues
